From afbe2997c8eda90669198f55d1afe983d9170b5e Mon Sep 17 00:00:00 2001
From: Pooja Prajod <a0132412@ti.com>
Date: Fri, 22 May 2015 16:31:23 +0530
Subject: [gst-ducati] gst-ducati: Change codec level param depending on
 stream level

The codec params were hard coded without check on stream caps.
For level 5 and 5.1 streams, presetLevelIdc should be set to
IH264VDEC_LEVEL51. This enables playing of level 5 streams.

Signed-off-by: Pooja Prajod <a0132412@ti.com>
---
 src/gstducatih264dec.c |   11 +++++++++++
 src/gstducatividdec.c  |    6 +++++-
 src/gstducatividdec.h  |    2 ++
 3 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/src/gstducatih264dec.c b/src/gstducatih264dec.c
index 962ceb4..e71264e 100644
--- a/src/gstducatih264dec.c
+++ b/src/gstducatih264dec.c
@@ -304,7 +304,18 @@ gst_ducati_h264dec_get_max_dpb_size (GstDucatiVidDec * self, GstCaps * caps)
 static gboolean
 gst_ducati_h264dec_set_sink_caps (GstDucatiVidDec * self, GstCaps * caps)
 {
+  GstDucatiH264Dec *h264dec = GST_DUCATIH264DEC (self);
+  IH264VDEC_Params *params = (IH264VDEC_Params *) self->params;
   GstStructure *structure;
+  const char *level;
+  structure = gst_caps_get_structure (caps, 0);
+  level = gst_structure_get_string (structure, "level");
+  GST_DEBUG_OBJECT (self, "Level obtained from stream caps = %s", level);
+
+  if (!strcmp (level, "5") || !strcmp (level, "5.1")) {
+    self->codec_create_params_changed = TRUE;
+    params->presetLevelIdc = IH264VDEC_LEVEL51;
+  }
 
   GST_DEBUG_OBJECT (self, "set_sink_caps: %" GST_PTR_FORMAT, caps);
 
diff --git a/src/gstducatividdec.c b/src/gstducatividdec.c
index 84bbf79..32dec0a 100644
--- a/src/gstducatividdec.c
+++ b/src/gstducatividdec.c
@@ -816,7 +816,9 @@ gst_ducati_viddec_parse_caps (GstDucatiVidDec * self, GstStructure * s)
      * need to re-create the codec:
      */
     if (G_UNLIKELY (self->codec)) {
-      if ((h != self->height) || (w != self->width)) {
+      if ((h != self->height) || (w != self->width)
+          || self->codec_create_params_changed) {
+        self->codec_create_params_changed = FALSE;
         codec_delete (self);
       }
     }
@@ -1816,6 +1818,8 @@ gst_ducati_viddec_init (GstDucatiVidDec * self, gpointer klass)
   self->ts_is_pts = FALSE;
 #endif
 
+  self->codec_create_params_changed = FALSE;
+
   self->pageMemType = XDM_MEMTYPE_TILEDPAGE;
 
   self->queried_external_pool = FALSE;
diff --git a/src/gstducatividdec.h b/src/gstducatividdec.h
index bc9f310..677b00a 100644
--- a/src/gstducatividdec.h
+++ b/src/gstducatividdec.h
@@ -124,6 +124,8 @@ struct _GstDucatiVidDec
 
   gboolean needs_flushing;
 
+  gboolean codec_create_params_changed;
+
   GHashTable *passed_in_bufs;
 
   GHashTable *dce_locked_bufs;
-- 
1.7.9.5

